lib/ui/page/home/page/my_profile/crop_avatar/widget/image_cropper/widget.dart

```diff
@@ -0,0 +1,448 @@
+        assert(touchSize > 0, 'touchSize cannot be zero'),
+        assert(gridThinWidth > 0, 'gridThinWidth cannot be zero'),
+        assert(gridThickWidth > 0, 'gridThickWidth cannot be zero'),
+        assert(minimumImageSize > 0, 'minimumImageSize cannot be zero');
+
+  @override
+  State<ImageCropper> createState() => _ImageCropperState();
+}
+
+class _ImageCropperState extends State<ImageCropper> {
+  /// [CropController] for managing the crop area and notifying listeners of changes.
+  late CropController controller;
+
+  /// Size of the image being displayed, which is updated dynamically with [onSize] function.
+  Size size = Size.zero;
+
+  /// The current crop handle point being interacted with, if any.
+  CropHandlePoint? cropHandlePoint;
+
+  /// Returns the positions of the crop handles in the current crop area.
+  /// It is accessed by [hitTest] to determine which handle is being interacted with.
+  Map<CropHandle, Offset> get cropHandlePositions => <CropHandle, Offset>{
+        CropHandle.upperLeft:
+            controller.crop.value.topLeft.scale(size.width, size.height),
+        CropHandle.upperRight:
+            controller.crop.value.topRight.scale(size.width, size.height),
+        CropHandle.lowerRight:
+            controller.crop.value.bottomRight.scale(size.width, size.height),
+        CropHandle.lowerLeft:
+            controller.crop.value.bottomLeft.scale(size.width, size.height),
+      };
+
+  @override
+  void initState() {
+    super.initState();
+    controller = widget.controller;
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+  }
+
+  /// Returns the ratio of the image's width to height.
+  double _getImageRatio() =>
+      controller.bitmapSize.width / controller.bitmapSize.height;
+
+  /// Returns the width of the image based on the rotation, maximum width and maximum height constraints.
+  double _getWidth(final double maxWidth, final double maxHeight) {
+    double imageRatio = _getImageRatio();
+    final screenRatio = maxWidth / maxHeight;
+    if (controller.rotation.value.isSideways) {
+      imageRatio = 1 / imageRatio;
+    }
+    if (imageRatio > screenRatio) {
+      return maxWidth;
+    }
+    return maxHeight * imageRatio;
+  }
+
+  /// Returns the height of the image based on the rotation, maximum width and maximum height constraints.
+  double _getHeight(final double maxWidth, final double maxHeight) {
+    double imageRatio = _getImageRatio();
+    final screenRatio = maxWidth / maxHeight;
+    if (controller.rotation.value.isSideways) {
+      imageRatio = 1 / imageRatio;
+    }
+    if (imageRatio < screenRatio) {
+      return maxHeight;
+    }
+    return maxWidth / imageRatio;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Center(
+      child: LayoutBuilder(
+        builder: (BuildContext context, BoxConstraints constraints) {
+          if (!controller.initialized) {
+            return const CircularProgressIndicator.adaptive();
+          }
```

Comment:
@SleepySquash  `GetxController`'s update method does the rerendering of this widget.

In this case, `update` method is called after initializing `bitmap` variable (It provides image dimension for rendering cropping square) within onInit function.

I chose to override initialize variable to check if bitmap is not equal to null. Rx variable wouldn't work here because I would call bitmap.value within the controller.  

Hence, I proceeded this way.  Do you suggest an alternative?

Reply:
@pratham-sarankar, I'm just wondering how this method causes the state to invoke the `build()` again. How it does that? I don't see a way for the `controller.update()` method to directly call `setState()` somewhere there.

Reply:
@SleepySquash 

According to `controller.update()`, method's documentation, 
>Rebuilds `GetBuilder` each time you call `update()`;

Internally, it calls `controller.refresh()`, which actually rebuilds the GetBuilder it is attached to.

Reply:
@pratham-sarankar, yep, `GetBuilder` does invoke `setState` under the hood. But which `GetBuilder` is the `controller` attached to?

Reply:
@SleepySquash,  `GetBuilder` is present in `lib/ui/page/home/page/my_profile/crop_avatar/view.dart`. 

```
  @override
  Widget build(BuildContext context) {
    final style = Theme.of(context).style;
    return GetBuilder(
      init: CropController(image: image, aspectRatio: 1),
      builder: (controller) {
        return Column(
           children: [
              ImageCropper(
                controller: controller,  
              ),
             // ... other widgets
           ]
        ); 
      },
    );
  }
```
It could be within `ImageCropper` widget but the rotation controls are in this widget.